### PR TITLE
Prevent race condition when closing client

### PR DESCRIPTION
**Pull request by lalexgap**
Fixes https://github.com/statechannels/go-nitro/issues/1330

This addresses a race condition that could occur when closing the nitro client. The nitro client starts a `handleEngineEvents` in a go-routine when it's constructed.  This functions listens on a chan from the engine for events and dispatches them to the `channelNotifier`.

The problem is that `handleEngineEvents` would to continue to run after we close the channel notifier in `Close`. This means `handleEngineEvents` would try to call `channelNotfier.Notify` after it's been closed, resulting in the error we see. Since the `toAPI` is [a buffered chan of size 100](https://github.com/statechannels/go-nitro/blob/4c820ee15dbf04648cb91e58617c75097cc1a26f/client/engine/engine.go#L142) it's quite possible that `handleEngineEvents` continues to execute for awhile after `Close` has been called.

To address this I've updated `handleEngineEvents` to accept a `context` so we can call `cancel` and stop the `handleEngineEvents` go-routine.

 While this let's us signal to the `handleEngineEvents` that it should stop, it's non-blocking, which means we can still end up closing the channel notifier before `handleEngineEvents` is done. To address this I've made use of `sync.Waitgroup` to block until the `handleEngineEvents` stops executing in `Close`. This way we can be sure that it's safe to close the channelNotifer.

I've run this on CI 8 times as a quick sanity check for the flicker. I did see one notification related failure but I think that is a [separate flicker](https://github.com/statechannels/go-nitro/issues/1323).

### CODE REVIEW

**Comment by lalexgap** [comment]
Closing in favour of #1332

**Comment by lalexgap** [comment]
> Closing in favour of #1332

Reopening as I think #1332 doesn't fix the problem.

File: client/client.go [hunk 1]
```diff
@@ -68,56 +73,62 @@ func New(messageService messageservice.MessageService, chainservice chainservice
...
-			close(d)
-
-			// use a nonblocking send to the RPC Client in case no one is listening
-			select {
-			case c.completedObjectivesForRPC <- completed.Id():
-			default:
+func (c *Client) handleEngineEvents(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
```
Comment (lalexgap) [inline client/client.go:97]: We actually don't have to use a `context` to signal `handleEngineEvents` to stop executing, we could just rely on the engine `ToAPI` chan getting closed when  we call `engine.Close` like we were doing previously.  

However I think this is less efficient as `close` has ["the effect of shutting down the channel after the last sent value is received"](https://pkg.go.dev/builtin#close), meaning this goroutine will only stop executing after it's gotten through the buffer of the `chan`.

By using a `context` and `cancel` the goroutine will stop as soon as it reads from `ctx.Done`

Comment (geoknee) [reply client/client.go:97]: I suppose the select statement could still end up "ignoring" the `ctx.Done()` path for some time as the buffer empties? It seems more likely that the goroutine is stopped sooner than via the existing behaviour, but I guess no certainty about that?

**Comment by geoknee** [comment]
> The problem is that `handleEngineEvents` would to continue to run after we close the channel notifier in `Close`

How about if we changed the order of operations in `client.Close()`, so that the `engine` is closed _before_ the client notifier? I think it makes sense to stop things that send before the things that receive from them?

File: client/client.go [hunk 2]
```diff
@@ -68,56 +73,62 @@ func New(messageService messageservice.MessageService, chainservice chainservice
...
 	c.channelNotifier = notifier.NewChannelNotifier(store, c.vm)
+
 	// Start the engine in a go routine
+
 	go c.engine.Run()
 
+	c.wg = &sync.WaitGroup{}
+	c.wg.Add(1)
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancelEventHandler = cancel
```
Comment (geoknee) [inline client/client.go:84]: When we resolve #1341, let's return to this section and add some comments to help the reader understand what is going on.

**Review by geoknee — APPROVED** [review]

**Review by geoknee — APPROVED** [review]

**Review by geoknee — APPROVED** [review]

**Pull request merged**